tools/codegen/api/types.py

```diff
@@ -211,6 +214,7 @@ def defn(self, name: Optional[str] = None, *, prefix: str = "") -> str:
     def _from_grouped_arguments(
         func: FunctionSchema,
         arguments: Sequence[Union[Argument, TensorOptionsArguments, ThisArgument]],
+        out_arguments: Sequence[Argument],
         *,
         faithful: bool
     ) -> 'CppSignature':
```

Comment:
Non-PR related philosophizing here: one of the consequences of immutable data types is that there is an equivalence between "data structure" and "memoized computation"; that is to say, if we have a constructor that does some nontrivial computation on data, we can choose to store the input data and then force clients to do the recompute every time, or we can upfront do the compute and then make it available. There isn't really any distinction and whether or not to slide from one end of the spectrum to the other is a mostly a matter of efficiency and clarity. (I guess this is why laziness is so good when you are purely functional).

Reply:
Another dimension that plays here is fanout - if there's a canonical representation and a bunch of useful projections, that often puts a natural boundary between what's stored and what's computed. 

Things get blurry when there's a good canonical representation, *but* a single useful projection kind of dominates. (I'm thinking of something like in the structured signature situation where most clients just want to iterate over some argument supertype, but the canonical representation would be much more precise, correct by construction etc.)

Argh SNIPED